src/main/java/org/casbin/jcasbin/main/SyncedEnforcer.java

```diff
@@ -95,22 +96,819 @@ public SyncedEnforcer(String modelPath, String policyFile, boolean enableLog) {
+     * If the rule already exists, the function returns false and the rule will not be added.
+     * Otherwise the function returns true by adding the new rule.
+     *
+     * @param ptype the policy type, can be "g", "g2", "g3", ..
+     * @param params the "g" policy rule.
+     * @return succeeds or not.
+     */
+    public boolean addNamedGroupingPolicy(String ptype, List<String> params) {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            return super.addNamedGroupingPolicy(ptype, params);
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * addNamedGroupingPolicy adds a named role inheritance rule to the current policy.
+     * If the rule already exists, the function returns false and the rule will not be added.
+     * Otherwise the function returns true by adding the new rule.
+     *
+     * @param ptype the policy type, can be "g", "g2", "g3", ..
+     * @param params the "g" policy rule.
+     * @return succeeds or not.
+     */
+    public boolean addNamedGroupingPolicy(String ptype, String... params) {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            return super.addNamedGroupingPolicy(ptype, params);
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * removeGroupingPolicy removes a role inheritance rule from the current policy.
+     *
+     * @param params the "g" policy rule, ptype "g" is implicitly used.
+     * @return succeeds or not.
+     */
+    public boolean removeGroupingPolicy(List<String> params) {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            return super.removeGroupingPolicy(params);
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * removeGroupingPolicy removes a role inheritance rule from the current policy.
+     *
+     * @param params the "g" policy rule, ptype "g" is implicitly used.
+     * @return succeeds or not.
+     */
+    public boolean removeGroupingPolicy(String... params) {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            return super.removeGroupingPolicy(params);
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * removeFilteredGroupingPolicy removes a role inheritance rule from the current policy, field filters can be specified.
+     *
+     * @param fieldIndex the policy rule's start index to be matched.
+     * @param fieldValues the field values to be matched, value ""
+     *                    means not to match this field.
+     * @return succeeds or not.
+     */
+    public boolean removeFilteredGroupingPolicy(int fieldIndex, String... fieldValues) {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            return super.removeFilteredGroupingPolicy(fieldIndex, fieldValues);
```

Comment:
It would lead to a deadlock. `super.removeFilteredNamedGroupingPolicy` will call overridden `super.removeFilteredNamedGroupingPolicy`.